pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java

```diff
@@ -48,8 +68,71 @@
 
     private final LinkedMap<TxnID, PositionImpl> aborts = new LinkedMap<>();
 
+    private final CompletableFuture<SystemTopicClient.Writer<TransactionBufferSnapshot>> takeSnapshotWriter;
+
+    // when add abort or change max read position, the count will +1. Take snapshot will set 0 into it.
+    private final AtomicLong changeMaxReadPositionAndAddAbortTimes = new AtomicLong();
+
+    private final Timer timer;
+
+    private final int takeSnapshotIntervalNumber;
+
+    private final int takeSnapshotIntervalTime;
+
     public TopicTransactionBuffer(PersistentTopic topic) {
+        super(State.None);
         this.topic = topic;
+        this.changeToInitializingState();
```

Comment:
what about  moving this initialisation procedure in a separate method ?

executing non trivial code in the constructor is problematic  because in case of problems you are not able to dispose correctly the object, because the caller does not receive a reference to the new object but we are still leaking references to the object.

Reply:
thanks. i will move it to recover method